### ISSUE TITLE
False positive: 'unsubscriptable-object'

### ISSUE THREAD

**Issue by tonial** [PyCQA/pylint#6033]
### Bug description

I have an instance of [`elasticsearch_dsl.search.Search` which has a `__getitem__` function](https://github.com/elastic/elasticsearch-dsl-py/blob/master/elasticsearch_dsl/search.py#L348).
Most other functions from `elasticsearch_dsl.search.Search` clone the instance before applying the new params, but pylint or astroid (I don't know which) thinks the cloned instance is of the parent class (which does not have the `__getitem__` function)

Here is a 'minimal' example without external dependencies :
```py
"""Module docstring"""


class ParentClass:  # pylint: disable=too-few-public-methods
    """Class docstring"""

    def clone(self):
        """function docstring"""
        copy = self.__class__()
        return copy


class MyClass(ParentClass):
    """Class docstring"""

    def __init__(self):
        """function docstring"""
        self.sort_keys = []
        self.extra = {}

    def clone(self):
        """function docstring"""
        copy = super().clone()
        copy.sort_keys = self.sort_keys
        copy.extra = self.extra
        return copy

    def sort(self, key):
        """function docstring"""
        copy = self.clone()
        copy.sort_keys.append(key)
        return copy

    def __getitem__(self, slicing):
        """function docstring"""
        copy = self.clone()
        copy.extra['from'] = slicing.start
        copy.extra['size'] = slicing.stop
        return copy


instance = MyClass().sort('timstamp')[:10]
print(instance.__class__)
print(instance.extra)
```

I guess that `self.__class__` in `ParentClass.clone()` is detected as `ParentClass` since in that function context, there's no way to know that is `self` is of a child class.

### Command used

```shell
`pylint myfile.py`
```

### Pylint output

```shell
************* Module pylint_test
pylint_test.py:31:8: E1101: Instance of 'ParentClass' has no 'sort_keys' member (no-member)
pylint_test.py:37:8: E1101: Instance of 'ParentClass' has no 'extra' member (no-member)
pylint_test.py:38:8: E1101: Instance of 'ParentClass' has no 'extra' member (no-member)
pylint_test.py:42:11: E1136: Value 'MyClass().sort('timstamp')' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

************* Module myfile

### Pylint version

```shell
pylint 2.13.2
astroid 2.11.2
Python 3.10.4 (main, Mar 24 2022, 16:12:56) [GCC 9.4.0]
```

### OS / Environment

Ubuntu 20.04

**Comment by jacobtylerwalls** [comment]
Thanks for the report. Sounds like another manifestation of #981 or PyCQA/astroid#457.

**Comment by jacobtylerwalls** [comment]
No worries! I'll mention this test case on both #981 and PyCQA/astroid#457 (and try to clarify those issue titles a bit)
